### ISSUE TITLE
[Question] Why does unscaling action behaves differently in training and eval

### ISSUE THREAD

**Issue by akane0314** [DLR-RM/stable-baselines3#1592]
### ❓ Question

I'm using PPO with `squash_output=True` option. It looks statistics (e.g., average reward, episode length) differs during collecting rollout and evaluating policy. After digging down, I found this is caused by the different behavior of unscaling action in `collect_rollouts` and  `predict` function used while train and eval, respectively. Why is `unscale_action` not applied in `collect_rollouts()`?

https://github.com/DLR-RM/stable-baselines3/blob/d68ff2e17f2f823e6f48d9eb9cee28ca563a2554/stable_baselines3/common/policies.py#L352-L361

https://github.com/DLR-RM/stable-baselines3/blob/d68ff2e17f2f823e6f48d9eb9cee28ca563a2554/stable_baselines3/common/on_policy_algorithm.py#L174-L177

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by akane0314** [comment]
Yes, I'm using custom env and boundaries are not in [-1, 1].

**Comment by MikhailGerasimov** [comment]
There is a related bug (or possibly a feature?) in SB3. It attempts to unsquash actions even when I use a distribution that does not support squashing. For example:
`model = PPO("MlpPolicy", env, policy_kwargs=dict(squash_output=True))`
Although my action space is bounded between -1 and 1, due to the unsquashing process, I receive values in the `env.step()` function that go beyond the boundaries of [-1, 1].

**Comment by araffin** [comment]
Hello,
I would welcome PR that solves both issues ;) (for the second one, we should not allow `squash_output=True` when not using gSDE).

**Comment by ReHoss** [comment]
Hi, 

Correct me if I am wrong but the "squash_output: Whether to squash the output using a tanh function" in the documentation is misleading as the $\tanh$ non-linearity when `squash_output=True` is only applied through `create_mlp` which is only called by `ContinuousCritic` and `Actor` objects in the source code, excluding for instance PPO. I rather observe a simpler scaling from the `self.unscale_actions` function.


If I understand correctly the first issue is the following:

- The `forward` method of the `nn.Module` does not unscale actions using action space bounds.
- The `predict` method of the `BasePolicy` **does unscale** actions using action space bounds.
- The rollout buffer contains actions that are not unscaled.
- The algorithm learns on the not unscaled actions.
- The `predict` method during evaluation implies interaction with the environments **using unscaled actions**


Consequently @araffin , what is the behavior you desire with respect to the `predict` function ?

Best,

**Comment by araffin** [comment]
> It looks like the squash_output argument is almost useless in the case of on-policy algorithms in the code and was developed for Experience Replay methods see create_mlp in Off-Policy methods?

Actually, the tanh comes from the distributions:
https://github.com/DLR-RM/stable-baselines3/blob/f4ec0f6afa1a23b0e0b746174cd0074471cc0b89/stable_baselines3/common/distributions.py#L250
and
https://github.com/DLR-RM/stable-baselines3/blob/f4ec0f6afa1a23b0e0b746174cd0074471cc0b89/stable_baselines3/common/distributions.py#L465-L468

the tanh in the `create_mlp` is for algorithms that don't rely on distributions (DDPG, TD3).

To sum up, everything which is internal uses scaled action, only the interaction with the env, so the `predict()` method unscales the action to the correct bounds when needed.
